File: android/core/src/main/kotlin/au/radar/core/Geo.kt
package au.radar.core

import kotlin.math.abs
import kotlin.math.asin
import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.min
import kotlin.math.sin
import kotlin.math.sqrt

/**
 * Spherical geometry, kept deliberately small. The same three functions exist in
 * RadarKit on iOS; both are checked against shared/alert-engine-fixtures.json.
 */
object Geo {
    private const val EARTH_RADIUS_M = 6_371_008.8

    private fun Double.toRadians() = this * Math.PI / 180.0

    private fun Double.toDegrees() = this * 180.0 / Math.PI

    /** Great-circle distance in metres. */
    fun distanceM(lat1: Double, lon1: Double, lat2: Double, lon2: Double): Double {
        val dLat = (lat2 - lat1).toRadians()
        val dLon = (lon2 - lon1).toRadians()
        val a = sin(dLat / 2) * sin(dLat / 2) +
            cos(lat1.toRadians()) * cos(lat2.toRadians()) * sin(dLon / 2) * sin(dLon / 2)
        return 2 * EARTH_RADIUS_M * asin(min(1.0, sqrt(a)))
    }

    /** Initial bearing from the first point to the second, degrees clockwise from north. */
    fun bearingDeg(lat1: Double, lon1: Double, lat2: Double, lon2: Double): Double {
        val dLon = (lon2 - lon1).toRadians()
        val y = sin(dLon) * cos(lat2.toRadians())
        val x = cos(lat1.toRadians()) * sin(lat2.toRadians()) -
            sin(lat1.toRadians()) * cos(lat2.toRadians()) * cos(dLon)
        return (atan2(y, x).toDegrees() + 360.0) % 360.0
    }

    /** The point a given distance along a bearing from a start point. */
    fun destination(lat: Double, lon: Double, bearingDeg: Double, distanceM: Double): Pair<Double, Double> {
        val angular = distanceM / EARTH_RADIUS_M
        val theta = bearingDeg.toRadians()
        val lat1 = lat.toRadians()
        val lon1 = lon.toRadians()
        val lat2 = asin(sin(lat1) * cos(angular) + cos(lat1) * sin(angular) * cos(theta))
        val lon2 = lon1 + atan2(
            sin(theta) * sin(angular) * cos(lat1),
            cos(angular) - sin(lat1) * sin(lat2),
        )
        return lat2.toDegrees() to lon2.toDegrees()
    }

    /** Smallest angle between two bearings, 0..180, wrapping across north. */
    fun bearingDelta(a: Double, b: Double): Double = abs(((a - b) % 360.0 + 540.0) % 360.0 - 180.0)
}

File: ios/RadarKit/Sources/RadarKit/Geo.swift
import Foundation

/// Spherical geometry, deliberately small. The same three functions exist in
/// `au.radar.core.Geo` on Android; both are checked against
/// `shared/alert-engine-fixtures.json`.
public enum Geo {
    static let earthRadiusM = 6_371_008.8

    /// Great-circle distance in metres.
    public static func distanceM(
        _ lat1: Double, _ lon1: Double,
        _ lat2: Double, _ lon2: Double
    ) -> Double {
        let dLat = (lat2 - lat1) * .pi / 180
        let dLon = (lon2 - lon1) * .pi / 180
        let a = sin(dLat / 2) * sin(dLat / 2)
            + cos(lat1 * .pi / 180) * cos(lat2 * .pi / 180) * sin(dLon / 2) * sin(dLon / 2)
        return 2 * earthRadiusM * asin(min(1, sqrt(a)))
    }

    /// Initial bearing from the first point to the second, degrees clockwise from north.
    public static func bearingDeg(
        _ lat1: Double, _ lon1: Double,
        _ lat2: Double, _ lon2: Double
    ) -> Double {
        let dLon = (lon2 - lon1) * .pi / 180
        let y = sin(dLon) * cos(lat2 * .pi / 180)
        let x = cos(lat1 * .pi / 180) * sin(lat2 * .pi / 180)
            - sin(lat1 * .pi / 180) * cos(lat2 * .pi / 180) * cos(dLon)
        return (atan2(y, x) * 180 / .pi + 360).truncatingRemainder(dividingBy: 360)
    }

    /// The point a given distance along a bearing from a start point.
    public static func destination(
        lat: Double, lon: Double, bearingDeg: Double, distanceM: Double
    ) -> (lat: Double, lon: Double) {
        let angular = distanceM / earthRadiusM
        let theta = bearingDeg * .pi / 180
        let lat1 = lat * .pi / 180
        let lon1 = lon * .pi / 180
        let lat2 = asin(sin(lat1) * cos(angular) + cos(lat1) * sin(angular) * cos(theta))
        let lon2 = lon1 + atan2(
            sin(theta) * sin(angular) * cos(lat1),
            cos(angular) - sin(lat1) * sin(lat2)
        )
        return (lat2 * 180 / .pi, lon2 * 180 / .pi)
    }

    /// Smallest angle between two bearings, 0...180, wrapping across north.
    public static func bearingDelta(_ a: Double, _ b: Double) -> Double {
        abs(((a - b).truncatingRemainder(dividingBy: 360) + 540)
            .truncatingRemainder(dividingBy: 360) - 180)
    }
}
